Fix memory stats for KVM

joseflauzino opened the pull request:
### Description

Using KVM hypervisor, the free memory stats of the user VMs that are returned by ACS do not correspond to those observed directly in the VMs (when using the `free -m` command, for example).

It was identified that the problem was in the memory stats collection process performed by Libvirt. This happens because the period that Libvirt would get memory stats updates for each VM was never set. Thus, even if ACS periodically requested the stats for the hosts, the data obtained were always the same.

This PR solves the mentioned problem. A new Agent configuration parameter called `vm.memballoon.stats.period` was created. This parameter allows operatos to set the time interval in which Libvirt will get memory stats updates. The default value of this parameter is 60 seconds, the same default value as the `vm.stats.interval` parameter. However, in a cloud containing multiple Management Servers, operators can set the value of `vm.memballoon.stats.period` lower than the `vm.stats.interval`, so that each Management Server always gets updated memory stats for all user VMs (since each Management Server can request stats at different instants).

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

### Feature/Enhancement Scale or Bug Severity

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [x] Minor
- [ ] Trivial

### How Has This Been Tested?

In a local lab, I tested multiple combinations for all the related Agent properties (`vm.memballoon.stats.period`, `vm.stats.interval`, and `vm.memballoon.disable`). In all cases tested, the memory stats of the user VMs were obtained properly.
Also, I added unit tests.

rohityadavcloud commented on agent/src/main/java/com/cloud/agent/Agent.java:
```diff
@@ -150,6 +153,29 @@ public int value() {
+     * Virsh command to get the memory balloon tag from the VM's XML file.<br><br>
+     * 1st parameter: the VM name;<br>
+     */
+    private static final String COMMAND_GET_MEM_BALLOON_FROM_XML_FILE = "virsh dumpxml %s | grep '<memballoon model='";
```
(comment for all virsh usages)
Is there a libvirt equivalent to do this? Forking to shell isn't a good idea with arbitrary strings `%s` that may be passed to it.

In case libvirt doesn't support that, you may also raise a MR with https://gitlab.com/libvirt/libvirt-java and then use those changes.

joseflauzino replied:
@rohityadavcloud Unfortunately, there are no corresponding methods for these specific virsh commands.
I think it would be better if we solved the collection of memory stats for now (even if it is using virsh commands). At another time we could add into libvirt-java corresponding methods for these and other commands already used in ACS.

joseflauzino commented:
@blueorangutan package

blueorangutan commented:
@joseflauzino a Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

ravening commented on agent/src/main/java/com/cloud/agent/Agent.java:
```diff
@@ -700,6 +728,62 @@ protected void processRequest(final Request request, final Link link) {
+    protected void setupMemoryBalloonStatsPeriod() {
+        String listAllVmsResult = Script.runSimpleBashScript(COMMAND_LIST_ALL_VMS);
+        if (StringUtils.isBlank(listAllVmsResult)) {
+            s_logger.debug("Skipping the memory balloon stats period setting, since there are no guest VMs.");
```
if period is set to 30 seconds, dont you think this line will populate the log file if there are no vm's?

joseflauzino commented:
I moved the code to a more appropriate location. Also, I adjusted the code to get the VM list (Libvirt domains) by using libvirt-java instead of a virsh command.
Thanks for the reviews.

blueorangutan commented:
@acs-robot a Jenkins job has been kicked to build UI QA env. I'll keep you posted as I make progress.

blueorangutan commented:
UI build: :heavy_check_mark:
Live QA URL: http://qa.cloudstack.cloud:8080/client/pr/6358 (SL-JID-1609)

DaanHoogland commented on agent/src/main/java/com/cloud/agent/properties/AgentProperties.java:
```diff
@@ -61,6 +61,21 @@
+     * Data type: Integer.<br>
+     * Default value: <code>60</code>
+     */
+    public static final Property<Integer> VM_MEMBALLOON_STATS_PERIOD = new Property<Integer>("vm.memballoon.stats.period", 60);
```
```suggestion
    public static final Property<Integer> VM_MEMBALLOON_STATS_PERIOD = new Property<>("vm.memballoon.stats.period", 60);
```

DaanHoogland reviewed (commented):
@joseflauzino ,
please have a look at the code smells reported:
> SonarCloud Quality Gate failed.    [![Quality Gate failed](https://camo.githubusercontent.com/4ea51c1f64ee3746f631653a02ab678ca6a3efb5f5cb474402faed2e3dcf90b5/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f5175616c6974794761746542616467652f6661696c65642d313670782e706e67)](https://sonarcloud.io/dashboard?id=apache_cloudstack&pullRequest=6358)
> 
> [![Bug](https://camo.githubusercontent.com/4c6102327f5a954f9c8acaf2e2714183157a9e41717b371b2cd585cf25057310/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f6275672d313670782e706e67)](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6358&resolved=false&types=BUG) [![A](https://camo.githubusercontent.com/1cba125a897d7fa47033a3b3b2be2bbee680d34d4f004a215564659b853fb201/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f526174696e6742616467652f412d313670782e706e67)](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6358&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6358&resolved=false&types=BUG) [![Vulnerability](https://camo.githubusercontent.com/3ba1ee49636ffc3427e38649a9f8a65ee392f28e8a662fcf96ce24cefbb520e9/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f76756c6e65726162696c6974792d313670782e706e67)](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6358&resolved=false&types=VULNERABILITY) [![A](https://camo.githubusercontent.com/1cba125a897d7fa47033a3b3b2be2bbee680d34d4f004a215564659b853fb201/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f526174696e6742616467652f412d313670782e706e67)](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6358&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6358&resolved=false&types=VULNERABILITY) [![Security Hotspot](https://camo.githubusercontent.com/fb735cbe76f8d5e1679c76ce83b740ceb1eaf62de4f7bf88623dc9953261aff7/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f73656375726974795f686f7473706f742d313670782e706e67)](https://sonarcloud.io/project/security_hotspots?id=apache_cloudstack&pullRequest=6358&resolved=false&types=SECURITY_HOTSPOT) [![A](https://camo.githubusercontent.com/1cba125a897d7fa47033a3b3b2be2bbee680d34d4f004a215564659b853fb201/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f526174696e6742616467652f412d313670782e706e67)](https://sonarcloud.io/project/security_hotspots?id=apache_cloudstack&pullRequest=6358&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=apache_cloudstack&pullRequest=6358&resolved=false&types=SECURITY_HOTSPOT) [![Code Smell](https://camo.githubusercontent.com/8fe18b2dfb6f7d4e44582f281b29f617eb5ae07c248d2002ca586e91da219212/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f636f64655f736d656c6c2d313670782e706e67)](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6358&resolved=false&types=CODE_SMELL) [![A](https://camo.githubusercontent.com/1cba125a897d7fa47033a3b3b2be2bbee680d34d4f004a215564659b853fb201/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f526174696e6742616467652f412d313670782e706e67)](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6358&resolved=false&types=CODE_SMELL) [12 Code Smells](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6358&resolved=false&types=CODE_SMELL)
> 
> [![71.6%](https://camo.githubusercontent.com/86b73734925524d51d47145ab1a1d3d06233748f7a9b0f77993d531ce5110d8e/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f436f76657261676543686172742f36302d313670782e706e67)](https://sonarcloud.io/component_measures?id=apache_cloudstack&pullRequest=6358&metric=new_coverage&view=list) [71.6% Coverage](https://sonarcloud.io/component_measures?id=apache_cloudstack&pullRequest=6358&metric=new_coverage&view=list) [![0.0%](https://camo.githubusercontent.com/8047c63e1f9ed03f63001e1eadce4676bade3e0f83ec690a9c625287796248a6/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f4475706c69636174696f6e732f332d313670782e706e67)](https://sonarcloud.io/component_measures?id=apache_cloudstack&pullRequest=6358&metric=new_duplicated_lines_density&view=list) [0.0% Duplication](https://sonarcloud.io/component_measures?id=apache_cloudstack&pullRequest=6358&metric=new_duplicated_lines_density&view=list)

some of them make more sense than others, but none arereally senseless as far as I can see.

blueorangutan commented:
Packaging result: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: debian :heavy_check_mark: suse15. SL-JID 3525

weizhouapache commented on agent/src/main/java/com/cloud/agent/properties/AgentProperties.java:
```diff
@@ -54,6 +54,28 @@
+     * Data type: boolean.<br>
+     * Default value: true.
+     */
+    public static final Property<Boolean> ENABLE_IO_URING = new Property<Boolean>("enable.io.uring", true);
```
@joseflauzino 
`ENABLE_IO_URING` is not needed I think.

weizhouapache commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java:
```diff
@@ -1288,9 +1301,67 @@ public boolean configure(final String name, final Map<String, Object> params) th
+
+        if (AgentPropertiesFileHandler.getPropertyValue(AgentProperties.VM_MEMBALLOON_DISABLE)) {
+            s_logger.debug(String.format("Skipping the memory balloon stats period setting because the [%s] property is set to 'true'.",
+                    AgentProperties.VM_MEMBALLOON_DISABLE.getName()));
```
what about the existing vms which has memory ballooning enabled ?
should memory ballooning be disabled ?

weizhouapache commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java:
```diff
@@ -1288,9 +1301,67 @@ public boolean configure(final String name, final Map<String, Object> params) th
+        Integer currentVmBalloonStatsPeriod = AgentPropertiesFileHandler.getPropertyValue(AgentProperties.VM_MEMBALLOON_STATS_PERIOD);
+
+        for (Integer vmId : vmIdList) {
+            String getMemBalloonFromXmlFileCommand = String.format(COMMAND_GET_MEM_BALLOON_FROM_XML_FILE, vmId);
```
it is better to update `LibvirtDomainXMLParser` to support `GuestResourceDef` , instead of 
`COMMAND_GET_MEM_BALLOON_FROM_XML_FILE = "virsh dumpxml %s | grep '<memballoon model='";`

weizhouapache commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtVMDef.java:
```diff
@@ -275,7 +279,14 @@ public String toString() {
+                memBalloonModel = "virtio";
+                statsPeriod = memoryBalloonStatsPeriod;
+            }
+            response.append(String.format("<devices>\n<memballoon model='%s'>\n<stats period='%s'/>\n</memballoon>\n</devices>\n", memBalloonModel, statsPeriod));
```
if model is 'none', we do not need to add `<stats period='xxx'/>`

rohityadavcloud commented:
ping @joseflauzino can you address the review comments?

joseflauzino commented:
@weizhouapache,  thanks for the review. I addressed the comments with the most straightforward solutions and will address the remaining ASAP.

GabrielBrascher commented on agent/conf/agent.properties:
```diff
@@ -202,6 +202,11 @@ hypervisor.type=kvm
+# The time interval (in seconds) at which the balloon driver will get memory stats updates.
+# This is equivalent to Libvirt's --period parameter when using the dommemstat command.
+# Default value: 60
+#vm.memballoon.stats.period=60
```
```suggestion
# vm.memballoon.stats.period=60
```

GabrielBrascher commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java:
```diff
@@ -440,6 +440,19 @@ public class LibvirtComputingResource extends ServerResourceBase implements Serv
+     * 1st parameter: the VM name;<br>
+     * 2nd parameter: the period (in seconds).
+     */
+    private static final String COMMAND_SET_MEM_BALLOON_STATS_PERIOD = "virsh dommemstat %s --period %s --live";
```
Why use `virsh` commands "by hand" if there is already Java bind for it via [Libvirt-Java Domain.memoryStats(int number)](https://gitlab.com/libvirt/libvirt-java/-/blob/master/src/main/java/org/libvirt/Domain.java#L1509)?
Is there something that the `memoryStats` does not support?

DaanHoogland replied:
Can you answer this one @joseflauzino , did you investigate the java binding?

GutoVeronezi replied:
ping @joseflauzino

joseflauzino replied:
@DaanHoogland and @GutoVeronezi 
I answered this one in [this other comment](https://github.com/apache/cloudstack/pull/6358#issuecomment-1212112323). That's why I marked it as resolved.

GabrielBrascher commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java:
```diff
@@ -440,6 +440,19 @@ public class LibvirtComputingResource extends ServerResourceBase implements Serv
+     * Virsh command to get the memory balloon tag from the VM's XML file.<br><br>
+     * 1st parameter: the VM name;<br>
+     */
+    private static final String COMMAND_GET_MEM_BALLOON_FROM_XML_FILE = "virsh dumpxml %s | grep '<memballoon model='";
```
Same here, there is already a Java bind for the `virsh dumpxml` via [Domain.getXMLDesc(int flags)](https://gitlab.com/libvirt/libvirt-java/-/blob/master/src/main/java/org/libvirt/Domain.java#L1296)

GabrielBrascher replied:
@joseflauzino there are some examples around the codebase. Maybe [this one](https://github.com/apache/cloudstack/blob/81b7e6e750a581375612b78fbf04ce94da573839/plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/storage/KVMStorageProcessor.java#L1184) can help :-)

joseflauzino commented:
Thanks for the review, @GabrielBrascher. I will address your comments as well. I will be pushing a new version of the code including all the revisions in the next few days.

joseflauzino commented:
Thanks for the review, @weizhouapache and @GabrielBrascher.

I have just committed the new version of the code. These are the relevant changes:

1. Now the stats parameter of the memballoon tag will be set to 0 when `vm.memballoon.disable=true`. 

2. Virsh commands have been replaced by methods from the Domain class. The only exception is the command to set the stats parameter of the memballoon tag (`virsh dommemstat <vm-name-or-id> --period <value-in-seconds> --live`), which cannot be replaced by the `memoryStats(int number)` method. The reason is that this method only allows us to get the memory stats and not to set the new gathering period in the VM XML (as desired).

blueorangutan commented:
@acs-robot a Jenkins job has been kicked to build UI QA env. I'll keep you posted as I make progress.

blueorangutan commented:
UI build: :heavy_check_mark:
Live QA URL: http://qa.cloudstack.cloud:8080/client/pr/6358 (SL-JID-2130)

joseflauzino commented:
@blueorangutan package

blueorangutan commented:
Packaging result: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: debian :heavy_check_mark: suse15. SL-JID 3964

shwstppr commented:
@blueorangutan package

blueorangutan commented:
@shwstppr a Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_multiplication_x: el7 :heavy_multiplication_x: el8 :heavy_multiplication_x: debian :heavy_multiplication_x: suse15. SL-JID 4003

blueorangutan commented:
Packaging result: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: debian :heavy_check_mark: suse15. SL-JID 4014

DaanHoogland commented:
@joseflauzino it is alright not to agree with each comment, but please don't resolve them unanswered. please answer them and leave it to the commenter to resolve them.

rohityadavcloud commented:
@blueorangutan package

blueorangutan commented:
@rohityadavcloud a Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_multiplication_x: el7 :heavy_multiplication_x: el8 :heavy_multiplication_x: debian :heavy_check_mark: suse15. SL-JID 4389

weizhouapache commented:
@blueorangutan package

blueorangutan commented:
@weizhouapache a Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: debian :heavy_check_mark: suse15. SL-JID 4519

DaanHoogland commented:
@blueorangutan test

blueorangutan commented:
@DaanHoogland a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

weizhouapache commented:
@blueorangutan test ubuntu22 kvm-ubuntu22

blueorangutan commented:
@weizhouapache a Trillian-Jenkins test job (ubuntu22 mgmt + kvm-ubuntu22) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian Build Failed (tid-5185)<b/>

blueorangutan commented:
<b>Trillian test result (tid-5184)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 39770 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr6358-t5184-kvm-centos7.zip
Smoke tests completed. 99 look OK, 5 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_add_primary_storage_disabled_host | `Error` | 0.60 | test_primary_storage.py
test_01_primary_storage_nfs | `Error` | 0.10 | test_primary_storage.py
ContextSuite context=TestStorageTags>:setup | `Error` | 0.18 | test_primary_storage.py
test_01_secure_vm_migration | `Error` | 156.36 | test_vm_life_cycle.py
test_02_unsecure_vm_migration | `Error` | 270.83 | test_vm_life_cycle.py
test_03_secured_to_nonsecured_vm_migration | `Error` | 141.97 | test_vm_life_cycle.py
test_08_migrate_vm | `Error` | 43.85 | test_vm_life_cycle.py
test_03_deploy_and_scale_kubernetes_cluster | `Failure` | 33.00 | test_kubernetes_clusters.py
test_07_deploy_kubernetes_ha_cluster | `Failure` | 60.73 | test_kubernetes_clusters.py
test_08_upgrade_kubernetes_ha_cluster | `Failure` | 38.07 | test_kubernetes_clusters.py
test_09_delete_kubernetes_ha_cluster | `Failure` | 39.99 | test_kubernetes_clusters.py
ContextSuite context=TestKubernetesCluster>:teardown | `Error` | 130.33 | test_kubernetes_clusters.py
test_02_list_snapshots_with_removed_data_store | `Error` | 8.56 | test_snapshots.py
test_02_list_snapshots_with_removed_data_store | `Error` | 8.56 | test_snapshots.py
test_hostha_enable_ha_when_host_in_maintenance | `Error` | 304.80 | test_hostha_kvm.py

DaanHoogland commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java:
```diff
@@ -1288,9 +1298,66 @@ public boolean configure(final String name, final Map<String, Object> params) th
+     * Sets the balloon driver of each VM to get the memory stats at the time interval defined in the agent.properties file.
+     * @param conn the Libvirt connection.
+     */
+    protected void setupMemoryBalloonStatsPeriod(Connect conn) {
```
50 line method. Can you disect it into smaller parts, please?

joseflauzino replied:
Done.

DaanHoogland commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtDomainXMLParser.java:
```diff
@@ -196,6 +198,18 @@ public boolean parseDomainXML(String domXML) {
                 diskDefs.add(def);
             }
 
+            NodeList memBalloons = devices.getElementsByTagName("memballoon");
+            if ((memBalloons != null) && (memBalloons.getLength() != 0)) {
+                Element memBalloon = (Element)memBalloons.item(0);
+                String model = memBalloon.getAttribute("model");
+                MemBalloonDef def = new MemBalloonDef();
+                if (model.equalsIgnoreCase("virtio")) {
+                    String statsPeriod = getAttrValue("stats", "period", memBalloon);
+                    def.defVirtioMemBalloon(statsPeriod);
+                }
+                memBalloonDef = def;
+            }
+
```
can this go in a separate method, please?

joseflauzino replied:
Done.

DaanHoogland commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtVMDef.java:
```diff
@@ -275,7 +280,14 @@ public String toString() {
             }
 
-            response.append(String.format("<devices>\n<memballoon model='%s'/>\n</devices>\n", this.memoryBalloning ? "virtio" : "none"));
+            MemBalloonDef memBalloonDef = new MemBalloonDef();
+            if (this.memoryBalloning) {
+                memBalloonDef.defVirtioMemBalloon(String.valueOf(memoryBalloonStatsPeriod));
+            } else {
+                memBalloonDef.defNoneMemBalloon();
+            }
+            response.append(String.format("<devices>\n%s\n</devices>\n", memBalloonDef.toString()));
+
```
can this go in a separate method, please?

joseflauzino replied:
I think there is no need in this case. I've only added a condition before appending the string. I also created a test case that validates this.

weizhouapache commented:
@joseflauzino 
I use some your changes in https://github.com/apache/cloudstack/pull/6571/commits/fb868986a290030332052561605a0903ac005fce

joseflauzino commented on agent/src/main/java/com/cloud/agent/properties/AgentProperties.java:
```diff
@@ -54,6 +54,21 @@
+     * Data type: Integer.<br>
+     * Default value: <code>60</code>
+     */
+    public static final Property<Integer> VM_MEMBALLOON_STATS_PERIOD = new Property<>("vm.memballoon.stats.period", 60);
```
Done.

joseflauzino commented:
@blueorangutan package

blueorangutan commented:
@joseflauzino a Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

weizhouapache commented on agent/conf/agent.properties:
```diff
@@ -202,6 +202,11 @@ hypervisor.type=kvm
 #
+# The time interval (in seconds) at which the balloon driver will get memory stats updates.
+# This is equivalent to Libvirt's --period parameter when using the dommemstat command.
+# Default value: 60
```
thanks @joseflauzino 
can you update the default value to 0 in  agent/conf/agent.properties ?

GutoVeronezi commented on agent/conf/agent.properties:
```diff
@@ -202,6 +202,11 @@ hypervisor.type=kvm
 #
+# The time interval (in seconds) at which the balloon driver will get memory stats updates.
+# This is equivalent to Libvirt's --period parameter when using the dommemstat command.
+# Default value: 0
```
As the commented property already has the default value, the line `# Default value: 0` is not necessary.

joseflauzino replied:
Done.

GutoVeronezi commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java:
```diff
@@ -1298,9 +1308,85 @@ public boolean configure(final String name, final Map<String, Object> params) th
+            return 0;
+        }
+        Integer vmBalloonStatsPeriod = AgentPropertiesFileHandler.getPropertyValue(AgentProperties.VM_MEMBALLOON_STATS_PERIOD);
+        if(vmBalloonStatsPeriod == 0) {
```
```suggestion
        if (vmBalloonStatsPeriod == 0) {
```

joseflauzino replied:
Done.

GutoVeronezi commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java:
```diff
@@ -1298,9 +1308,85 @@ public boolean configure(final String name, final Map<String, Object> params) th
+                dm = conn.domainLookupByID(vmId);
+                parser.parseDomainXML(dm.getXMLDesc(0));
+                MemBalloonDef memBalloon = parser.getMemBalloon();
+                if (MemBalloonDef.MemBalloonModel.VIRTIO.equals(memBalloon.getMemBalloonModel())) {
+                    String setMemBalloonStatsPeriodCommand = String.format(COMMAND_SET_MEM_BALLOON_STATS_PERIOD, vmId, currentVmBalloonStatsPeriod);
+                    String setMemBalloonStatsPeriodResult = Script.runSimpleBashScript(setMemBalloonStatsPeriodCommand);
+                    if (StringUtils.isNotBlank(setMemBalloonStatsPeriodResult)) {
+                        s_logger.error(String.format("Unable to set up memory balloon stats period for VM (Libvirt Domain) with ID [%s] due to an error when running the [%s] "
+                                + "command. Output: [%s].", vmId, setMemBalloonStatsPeriodCommand, setMemBalloonStatsPeriodResult));
+                        continue;
+                    }
+                    s_logger.debug(String.format("The memory balloon stats period [%s] has been set successfully for the VM (Libvirt Domain) with ID [%s] and name [%s].",
+                            currentVmBalloonStatsPeriod, vmId, dm.getName()));
+                    continue;
+                }
+                s_logger.debug(String.format("Skipping the memory balloon stats period setting for the VM (Libvirt Domain) with ID [%s] and name [%s] because this VM has no memory"
+                        + " balloon.", vmId, dm.getName()));
```
We could invert those statements to avoid hadouken ifs.

joseflauzino replied:
Done.

GutoVeronezi commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtDomainXMLParser.java:
```diff
@@ -342,6 +346,25 @@ public boolean parseDomainXML(String domXML) {
+    private MemBalloonDef parseMemBalloonTag(Element devices) {
+        MemBalloonDef def = new MemBalloonDef();
+        NodeList memBalloons = devices.getElementsByTagName("memballoon");
+        if ((memBalloons != null) && (memBalloons.getLength() != 0)) {
```
```suggestion
        if (memBalloons != null && memBalloons.getLength() != 0) {
```

joseflauzino replied:
Done.

joseflauzino commented:
@blueorangutan package

blueorangutan commented:
@joseflauzino a Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: debian :heavy_check_mark: suse15. LL-JID 212

blueorangutan commented:
@DaanHoogland a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-46)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 44501 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr6358-t46-kvm-centos7.zip
Smoke tests completed. 102 look OK, 2 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_02_upgrade_kubernetes_cluster | `Failure` | 264.05 | test_kubernetes_clusters.py
test_08_upgrade_kubernetes_ha_cluster | `Failure` | 492.99 | test_kubernetes_clusters.py
test_01_redundant_vpc_site2site_vpn | `Failure` | 749.85 | test_vpc_vpn.py
test_01_vpc_site2site_vpn_multiple_options | `Error` | 1275.47 | test_vpc_vpn.py
test_01_vpc_site2site_vpn | `Error` | 546.30 | test_vpc_vpn.py
